File: CIF_GET/app/src/main/java/com/example/cif_get/CartActivity.java
package com.example.cif_get;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.ListView;
import android.widget.TextView;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class CartActivity extends AppCompatActivity {

    private ListView cartListView;
    private TextView totalPriceTextView;
    private Button checkoutButton;
    private Button delete, add;

    FirebaseFirestore db;

    private ArrayList<ModelClass> cartItems;
    private double totalPrice;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_cart);

        db = FirebaseFirestore.getInstance();

        ModelClass cart = (ModelClass) getIntent().getSerializableExtra("item");

        totalPriceTextView = findViewById(R.id.totalPriceTextView);
        checkoutButton = findViewById(R.id.checkoutButton);
        cartListView = findViewById(R.id.cartListView);

        totalPrice = 0.0;

        cartItems = new ArrayList<>();

        if (cart != null && cart.getId() != null) {
            // Check if the item already exists in the database
            db.collection("cartList")
                    .whereEqualTo("id", cart.getId())
                    .get()
                    .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                        @Override
                        public void onComplete(@NonNull Task<QuerySnapshot> task) {
                            if (task.isSuccessful()) {
                                if (task.getResult().isEmpty()) {
                                    // Item doesn't exist, add it to the database
                                    cart.setCount(1);
                                    db.collection("cartList")
                                            .add(cart)
                                            .addOnSuccessListener(new OnSuccessListener<DocumentReference>() {
                                                @Override
                                                public void onSuccess(DocumentReference documentReference) {
                                                    Log.d("Firestore", "Adding item");
                                                    retrieveCartItems();
                                                }
                                            });
                                } else {
                                    DocumentSnapshot documentSnapshot = task.getResult().getDocuments().get(0);
                                    String documentId = documentSnapshot.getId();

                                    int currentCount = documentSnapshot.getLong("count").intValue();
                                    int updatedCount = currentCount + 1;

                                    Map<String, Object> data = new HashMap<>();
                                    data.put("count", updatedCount);

                                    db.collection("cartList")
                                            .document(documentId)
                                            .update(data)
                                            .addOnSuccessListener(new OnSuccessListener<Void>() {
                                                @Override
                                                public void onSuccess(Void unused) {
                                                    Log.d("Firestore", "Count updated for duplicate item");
                                                    retrieveCartItems();
                                                }
                                            });
                                }
                            } else {
                                Log.e("Firestore", "Error checking item existence", task.getException());
                            }
                        }
                    });
        } else {
            retrieveCartItems();
        }
    }

    private void retrieveCartItems() {
        db.collection("cartList")
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()) {
                            cartItems.clear();
                            totalPrice = 0.0;

                            for (QueryDocumentSnapshot document : task.getResult()) {
                                ModelClass cartItem = document.toObject(ModelClass.class);
                                cartItems.add(cartItem);

                                String itemPricedollar = cartItem.getPrice();
                                String itemPrice = itemPricedollar.substring(1, itemPricedollar.length());
                                totalPrice = totalPrice + Double.parseDouble(itemPrice) * (cartItem.getCount());
                            }

                            totalPriceTextView.setText(String.valueOf(totalPrice));
                            updateListView();
                        } else {
                            Log.e("Firestore", "Error retrieving cart items", task.getException());
                        }
                    }
                });
    }

    private void updateListView() {
        CartAdapter adapter = new CartAdapter(CartActivity.this, cartItems, totalPrice, totalPriceTextView);
        cartListView.setAdapter(adapter);
        adapter.notifyDataSetChanged();
    }
}

File: QRBUY/app/src/main/java/com/example/qrbuy/CartActivity.kt
package com.example.qrbuy

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.ImageButton
import android.widget.ListView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.QuerySnapshot
import java.text.SimpleDateFormat
import java.time.LocalDateTime
import java.util.Date
import kotlin.random.Random

class CartActivity : AppCompatActivity() {

    private val cartItems: ArrayList<ModelClass> = ArrayList()
    private lateinit var listView: ListView
    private lateinit var totalprice: TextView
    private lateinit var checkoutButton: Button
    private lateinit var backButton: ImageButton

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_cart)

        backButton = findViewById(R.id.backBtn)
        listView = findViewById(R.id.cartListView)
        totalprice = findViewById(R.id.totalPriceTextView)
        checkoutButton = findViewById(R.id.checkoutButton)

        val UserID = intent.getStringExtra("UserID") ?: ""

        backButton.setOnClickListener{
            finish()
        }

        retrieveCartItems(UserID)

    }

    private fun retrieveCartItems(UserID: String) {
        var total: Float = 0.0f
        val db = FirebaseFirestore.getInstance()
        val DBlocation = db.collection("Users").document(UserID).collection("cartItems")
        DBlocation
            .get()
            .addOnSuccessListener { querySnapshot: QuerySnapshot ->
                for (document in querySnapshot) {
                    val item = document.toObject(ModelClass::class.java)
                    total += item.price ?: 0.0f
                    cartItems.add(item)
                }
                displayCartItems()
                recalculateTotalPrice(total)
//                totalprice.text = "$"+total.toString()

                checkoutButton.setOnClickListener {
                    copyCartItemsToTransaction(UserID)
                }
            }
            .addOnFailureListener { e ->
                Log.w("firebase fail", "Error retrieving items", e)
            }
    }

    private fun displayCartItems() {
        val adapter = CartItemAdapter(this, cartItems, totalprice)
        listView.adapter = adapter
        adapter.notifyDataSetChanged()
    }

    private fun recalculateTotalPrice(total: Float) {
        totalprice.text = "$" + total.toString()
    }

    private fun copyCartItemsToTransaction(UserID: String) {
        val db = FirebaseFirestore.getInstance()
        val cartItemsRef = db.collection("Users").document(UserID).collection("cartItems")
        val datetime = getCurrentDateTime()
        val rand = Random.nextInt(10000000, 100000000).toString()
        val transactionRef = db.collection("Users").document(UserID).collection("transaction").document(rand)

        cartItemsRef.get()
            .addOnSuccessListener { querySnapshot: QuerySnapshot ->
                val batch = db.batch()

                for (document in querySnapshot) {
                    val item = document.toObject(ModelClass::class.java)
                    item.transactionDate=datetime
                    item.transactionID=rand
                    val newItemRef = transactionRef.collection("items").document(document.id)
                    batch.set(newItemRef, item)
                }

                batch.commit()
                    .addOnSuccessListener {
                        Log.d("firebase success", "Cart items copied to transaction successfully.")

                        // Store the rand value in the "transactionNames" collection
                        val transactionNamesRef = db.collection("Users").document(UserID).collection("transactionNames").document(rand)
                        val data = hashMapOf<String, Any>("datetime" to datetime)
                        transactionNamesRef.set(data)
                            .addOnSuccessListener {
                                Log.d("firebase success", "Rand value stored in transactionNames collection.")
                                val intent = Intent(this, PaymentActivity::class.java)
                                intent.putExtra("UserId", UserID)
                                intent.putExtra("TransactionId", rand)
                                intent.putExtra("TransactionDate", datetime)
                                startActivity(intent)
                            }
                            .addOnFailureListener { e ->
                                Log.w("firebase fail", "Error storing rand value in transactionNames collection", e)
                            }

                        // Perform any necessary actions after copying the items
                    }
                    .addOnFailureListener { e ->
                        Log.w("firebase fail", "Error copying cart items to transaction", e)
                    }
            }
            .addOnFailureListener { e ->
                Log.w("firebase fail", "Error retrieving cart items", e)
            }
    }


    private fun getCurrentDateTime(): String {
        val currentDate = Date()

        // Format the date and time
        val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        val formattedDate = dateFormat.format(currentDate)

        return formattedDate
    }

}
